### DISCUSSION TITLE
Generation of plot for mesh containing polyhedron cells and hexahedron cells

### DISCUSSION THREAD

**Discussion by Saivellampati** [pyvista/pyvista#1835] [Q&A]
Hello Everyone.,

I have a cae file that I am trying to convert to vtk_unstructured using pyvista for visualization of mesh. That cae file has polyhedron and hexahedron cells.  I followed a usual process for converting to unstructured grid that requires cells, points, celltypes. But mesh is not rendering. This is happening for all cae files which are having polyhedron cells only. Is there any different process available for polyhedron cells?. can anyone help me for this issue.

**Comment by banesullivan** [comment]
I suspect something is wrong in your generation of the mesh, can you share your file and your script?

----

Polyhedron and hexahedron cell types are well supported as PyVista's rendering is entirely handled by VTK. Here is an example with  `vtkTriQuadraticHexahedron ` cells:

```py
from pyvista import examples

mesh = examples.download_tri_quadratic_hexahedron()
```

```py
>>> mesh.GetCell(0)
<vtkmodules.vtkCommonDataModel.vtkTriQuadraticHexahedron(0x6000009595f0) at 0x7f92b8e77e80>
```

```py
mesh.plot(show_edges=True, notebook=0, color=True)
```

<img width="624" alt="Screen Shot 2021-11-30 at 11 42 54 PM" src="https://user-images.githubusercontent.com/22067021/144185087-507c89bf-53a2-4f2f-982e-2a2c090a730d.png">

**Reply by Saivellampati** [reply]
Hi @banesullivan 
         Due to privacy reasons I am not sharing my code. I am providing two python scripts to explain my problem. I generated a mesh (image is attached) which contains only one polyhedron cell using vtk, that is working fine. My code is provided below

```py
import numpy as np
import pyvista as pv
import vtk
import vtk.util.numpy_support as np_support
_ugrid = vtk.vtkUnstructuredGrid()
_offsets = np.array([0, 51])
_cell_connectivity = np.array([10, 4, 0,1,2,3,4, 3,4,7,2,4, 7,6,1,2,4, 6,7,11,10,4, 10,11,8,9,4, 11,7,4,8,4, 8,9,5,4,4, 3,4,5,0,4, 0,5,6,1, 4, 6,5,9,10 ])
_coordinates = np.array(
    [
        [0, 0, 0],
        [0, 1, 0],
        [0, 1, 1],
        [0, 0, 1],
        [0.5, 0, 1],
        [0.5, 0, 0],
        [0.5, 1, 0],
        [0.5, 1, 1],
        [1, 0, 1],
        [1, 0, 0],
        [1, 1, 0],
        [1, 1, 1]   
    ])
vtk_coordinates = vtk.vtkPoints()
vtk_points = np_support.numpy_to_vtk(np.array(_coordinates, dtype=np.float32))
vtk_coordinates.SetNumberOfPoints(_coordinates.shape[0])
vtk_coordinates.SetData(vtk_points)

_ugrid.SetPoints(vtk_coordinates)

_number_of_cells = _offsets.shape[0] - 1
cell_numpy_type = np.empty(_number_of_cells, dtype=np.int32)
cell_numpy_type.fill(vtk.VTK_POLYHEDRON)
  
cell_type = np_support.numpy_to_vtk(cell_numpy_type,
                                            array_type=vtk.vtkUnsignedCharArray().GetDataType())                                  
cell_values = np_support.numpy_to_vtk(_cell_connectivity,
                                              array_type=vtk.vtkIdTypeArray().GetDataType())
        
face_offsets = np_support.numpy_to_vtk(_offsets,
                                               array_type=vtk.vtkIdTypeArray().GetDataType())
                                              
cell_array = vtk.vtkCellArray()
cell_array.SetNumberOfCells(_number_of_cells)
cell_array.SetData(face_offsets, cell_values)
_ugrid.SetCells(cell_type, cell_array)
grid = pv.UnstructuredGrid(_ugrid)
```


  I tried same with pyvista. But that is not working. That part of the code is below.

```py
import pyvista as pv
import vtk
import numpy as np
cells = np.array([12,0,1,2,3,4,5,6,7,8,9,10,11])
cell_type = np.array([vtk.VTK_POLYHEDRON])
points = np.array(
    [
        [0, 0, 0],
        [0, 1, 0],
        [0, 1, 1],
        [0, 0, 1],
        [0.5, 0, 1],
        [0.5, 0, 0],
        [0.5, 1, 0],
        [0.5, 1, 1],
        [1, 0, 1],
        [1, 0, 0],
        [1, 1, 0],
        [1, 1, 1]   
    ])

grid = pv.UnstructuredGrid(cells, cell_type, points)
_ = grid.plot(show_edges=True)
```
 please have look at both the codes and tell me where I am doing wrong in case pyvista.



![polyhedron](https://user-images.githubusercontent.com/87307306/145075611-ef1a2c17-b530-4dc0-a855-b10dab3d5cb3.png)
